In Image control, return the image's ID from block_value() when $echo is not true

kienstra opened the pull request:
* The behavior shouldn't change for `block_field( 'foo-image-field' )` or `block_value( 'foo-image-field, false )`
* But for `block_value()`, this should make available the attachment ID
* It looks like many functions for attachments take the ID as a parameter, not the `WP_Post`
* Like `wp_get_attachment_image()` and `wp_get_attachment_image_sizes()`

Closes #256. Closes #247.

kienstra committed:
Begin work to return attachment WP_Post for block_value( 'foo-image-field' )

The behavior shouldn't change for block_field( 'foo-image-field' )
But for block_value() without a 2nd parameter of true (to echo),
this should make available the attachment WP_Post object.

kienstra committed:
Instead of exposing the WP_Post object, make available the ID

This WP_Post doesn't look to have much information.
But the ID is useful in functions like
wp_get_attachment_image()
wp_get_attachment_image_sizes()

kienstra committed:
Get the image's ID to output for the Image control

Though the image's URL is stored,
get the ID to output via block_value( 'foo-image' ).

kienstra commented on php/blocks/controls/class-image.php:
```diff
@@ -57,4 +57,68 @@ public function register_settings() {
+				'post_type'   => 'attachment',
+				'post_status' => 'inherit',
+				'fields'      => 'ids',
+				'meta_query'  => array( // phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_query
```
Eventually, we might consider storing the ID of the image in the block, in addition to the URL. Right now, it only stores the URL.

This might help avoid running a meta query to get the ID.

kienstra replied:
@todo: consider using `attachment_url_to_postid()` instead of this helper method.

kienstra commented:
This is almost ready for review, pending a little testing on my part.

kienstra committed:
Use a Core function to get the attachment's ID

Instead of the custom method,
use the core function attachment_url_to_postid()
If it returns 0, return false from validate().

kienstra commented:
**Expected Results For Image Control**

| Function call | Expected result  |  Notes  | 
|---|---|---|
| `block_value( 'foo-valid-url' )`       | returns attachment `ID`   | Different from before, used to return the URL  |
| `block_value( 'foo-invalid-url' )`       | returns `0` | Also different from before. Returns `false` if the URL isn't on the same site, or it doesn't match an existing attachment (image). |
| `block_field( 'foo-valid-or-invalid-url' )`       | echoes the string URL | No validation added in this PR,  simply echoes the string (no change from previous behavior) |   

Updated: `block_value( 'foo-invalid-url' )`  now returns `0`

kienstra commented:
**Testing Steps**

1. Create a block with an Image control, or use an existing one:
<img width="1006" alt="block-image-control" src="https://user-images.githubusercontent.com/4063887/54948622-be22ee80-4f02-11e9-9fae-5ec3fde98fa1.png">

2. Create a new post with that block, and add an image from the same site. External images, like from http://lorempixel.com, will still work with `block_field()`. But `block_value()` won't get the image ID, as they're not from this site.
3. Add a template like `blocks/block-test-image.php`:

```php
<?php
/**
 * Template for test-image block to test the output of the Image control
 */
$field_name = 'test-image';

?>
<p>There should be an image below:</p>
<img src="<?php block_field( $field_name ); ?>">

<p>Here is the result of block_value(), which should be an int:</p>
<?php echo block_value( $field_name ); ?>

<p>And here is the result of calling wp_get_attachment_image() with the result of block_value(). This should be an image:</p>
<?php echo wp_get_attachment_image( block_value( $field_name ) ); ?>

<p>Here is the result of block_value with an invalid field, this should be false an nothing should display below:</p>
<?php wp_get_attachment_image( block_value( 'random-invalid-field-name' ) ); ?>

```

Here's the expected front-end display of that template:

![expected-block-lab](https://user-images.githubusercontent.com/4063887/54948708-edd1f680-4f02-11e9-99d7-5afc3005b09c.png)

kienstra committed:
Fix the @return tag to be more specific

Before, this was simply mixed,
but that doesn't help much.

kienstra commented:
**Request For Review**

Hi @lukecarbis and @RobStino,
Could you please review this when you have a chance? There are test steps and expected behavior above.

Thanks!

lukecarbis commented:
Closes #256.

lukecarbis commented on php/blocks/controls/class-image.php:
```diff
@@ -57,4 +57,20 @@ public function register_settings() {
+		if ( $echo ) {
+			return $value;
+		} else {
+			$id = attachment_url_to_postid( $value );
```
What an awesome function!

kienstra replied:
Haha, I just found out about it

lukecarbis commented:
This PR works for me as is, and I'd be happy to merge it if everyone else is comfortable, but there's something nagging at me with this approach.

Are we allowing the editor to paste in any old image URL? If the image is not in the media library, this could have unexpected results in the template, as a `false` response would be returned even for valid images.

I propose that we suggest block creators use a "URL" field if they want to support off-site images, or a copy / paste workflow.

For the image field, we should _force_ the media library to be used, to provide block creators as much consistency as possible. In this case we would:

- Removing the text input containing the URL when editing a block
- Provide a preview image when editing a block instead
- Store the Attachment Post ID in the database, instead of the URL
- Fetch the URL _from the ID_ when using `block_field`

This approach will also be more consistent with the User, Post, and Taxonomy fields storing IDs in the database.

RobStino commented:
Shall I hold off the review until the above is decided upon?

kienstra commented:
**ID or URL?**

Hi @lukecarbis,
Good points.

>Are we allowing the editor to paste in any old image URL?

Yes, there's no validation now if the URL is hosted on the site, when using `block_field( 'foo-image-field' )`. Though there is when using `block_value( 'foo-image-field' )`, as it then tries to get the image's `ID`.

Right, I think the URL control would be more appropriate for external images not hosted on the site. I think all 4 of your points above are right.

I could work on converting the existing URL strings to attachment IDs. Of course, there would be a backward-compatibility break if users have external URLs in the Image control.

kienstra commented:
Hi @RobStino,
>Shall I hold off the review until the above is decided upon?

Yes, this PR might change a lot.

lukecarbis commented:
@kienstra I think go for the backwards compatibility break – we'll make sure that we communicate that to our users before the next release.

kienstra commented:
@lukecarbis, sounds good.

kienstra committed:
Change Image control to show image in editor

Remove the <input>, and instead show the image.
There's work remaining, including styling
of the 'Media Library' and 'Upload' buttons.

lukecarbis commented on js/blocks/controls/image.js:
```diff
@@ -1,90 +1,111 @@
+	} )( ( ownProps) => {
+		const { imageAlt, imageSrc, setAttributes } = ownProps;
+		const attr = { ...ownProps.attributes };
+		const isUploading = 'undefined' !== typeof attr[ field.name ] && 'string' === typeof attr[ field.name ] && 'Uploading' === attr[ field.name ].substr( 0, 9 );
```
I tried for ages to find a way of setting `isUploading` is a property of the BlockLabImageControl, so it could be set dynamically instead of just relying on the value of the field. I think there's a better way here, but I don't know it. If you don't know it either, happy to commit this change.

kienstra replied:
Sure, I'll look at this a little more. It looked pretty complex setting an `isUploading` property 😄

kienstra replied:
https://github.com/getblocklab/block-lab/pull/260/commits/dd5ad114c17144f630a1fa4b2640f266d8d982e4 introduces a state of `isUploading`. It seems to work so far.

![uploading-image-control](https://user-images.githubusercontent.com/4063887/55473835-e37fbe80-55cc-11e9-9fd7-9a204c4d4fdf.gif)

lukecarbis commented on php/blocks/controls/class-image.php:
```diff
@@ -46,15 +53,26 @@ public function register_settings() {
+	public function validate( $value, $echo ) {
+		// Backwards compatibility, as the value used to be the image's URL instead of its post ID.
+		if ( empty( intval( $value ) ) ) {
+			$value = attachment_url_to_postid( $value );
```
NICE! This means there's no back-compat break, right?

kienstra replied:
Yeah, this should minimize the back-compat break. If someone had entered an external URL in the `<input>`, this wouldn't help. But it'll usually find the ID for a URL that's from the Media Library.

And as you know, the output of `block_value( 'foo-image-field' )` is changing to output an ID instead of the URL string.

kienstra replied:
[Below](https://github.com/getblocklab/block-lab/pull/260#issuecomment-479475773) are more details on backwards-compatibility.

kienstra committed:
Use withState for the loading state

When the image is uploading,
set a state of isUploading.

kienstra committed:
Improve styling of 'Upload' and 'Media Library' buttons

Place them below the image,
and alongside each other.

kienstra committed:
If the Image control value is a string, use it as the image's src

This is for backwards-compatibility.
Before, the value was a string.

kienstra committed:
Add a comment to describe the Backward-compatibility logic

kienstra committed:
Add a button to remove the image

Before, there was an <input> for the image URL,
so it was possible to remove an image.
So add a way to remove an image.

kienstra committed:
Change the text that displays if there's no image

Before, it was like field.help text.
Instead, state that there is no image selected.

RobStino commented:
Done some testing.

- Created a new block
- Added an image field for the editor and one for the inspector
- Added the below to my template file

````
<img src="<?php block_field( 'image-editor' ); ?>" alt="">
<img src="<?php block_field( 'image-inspector' ); ?>" alt="">
<?php $image_editor = block_value( 'image-editor' ); ?>
<?php $image_inspector = block_value( 'image-inspector' ); ?>
<?php echo $image_editor; ?>
<?php echo $image_inspector; ?>
````

I then observed the following in the editor preview
![image](https://user-images.githubusercontent.com/1578148/55477866-8a13a000-565d-11e9-871e-dbc3e98fe516.png)

And saw the following on the front end with 2017 theme.
![image](https://user-images.githubusercontent.com/1578148/55477994-d2cb5900-565d-11e9-9d4c-25c34bb4f73b.png)


Noted
- block_value() gives me the ID. This is now expected correct?
- block_field() returns the thumbnail URL rather than the original. Is this expected?

kienstra committed:
Improve backward-compatibility to allow for external images

In the previous <input>, you could add anything.
So this will use that external image

kienstra committed:
Add assertions for an external URL

The validate() method should still return an external URL
if $echo is true.

kienstra commented:
Hi @RobStino,
Thanks for testing this.

> block_value() gives me the ID. This is now expected correct?

Yes

> block_field() returns the thumbnail URL rather than the original. Is this expected?

Good point, https://github.com/getblocklab/block-lab/pull/260/commits/bc589e5d42c0d5563c679488fb41da9c48de7054 outputs the full image size.

kienstra commented:
**Backwards Compatibility**

The expected behavior for this is to have full backward compatibility for `block_field( 'foo' )`, for a URL that was saved in the previous `<input>`. Even with external URLs.

`block_value( 'foo-image' )` is the main difference in this PR. If the user had already saved a URL with the previous `<input>`, this will look for an ID to return for it. In my testing, that's worked well so far.

But for external images, it won't find an ID, and will return `0`.

Also, the behavior for when it doesn't find an ID is changed. It now returns `0` instead of `false`.

kienstra committed:
Change an assertion to expect 0 instead of false

The new return of validat() should be 0
when it cannot get the ID.

kienstra commented:
**Request For Review**

Hi @lukecarbis,
Could you please review this when you can?

`block_field()` should be backwards-compatible. Though `block_value()` now outputs the ID, and `0` for external images. 

So far, it looks like `block_value()` has found the `ID` for Media Library images that were added before this PR.

Also, I added a 'Remove' button if that's alright. Before, you could simply delete the URL from the `<input>`

![remove-button](https://user-images.githubusercontent.com/4063887/55480561-d5866980-55dd-11e9-8c18-92eafab085f6.gif)

Commit:
Update description when no image exists

Commit:
Update styles for media control (and remove unused styles)

lukecarbis reviewed (changes requested):
@kienstra Changes look great. I made a few minor style changes as well.

I noticed a bug, though. Steps to reproduce: 

1. Add a block with a media control
2. Add an image to the block
3. Blur focus from the block to preview (everything looks good)
4. Click the block to edit it again
5. Remove the image
6. Blur focus from the block to preview

I get this error:
<img width="688" alt="Screen Shot 2019-04-04 at 8 05 43 am" src="https://user-images.githubusercontent.com/1097667/55516505-b86e9b00-56b0-11e9-9e15-3d1b7ed1b10f.png">

The Remove button is a great idea, by the way! Code looks great too.

kienstra commented:
Hi @lukecarbis, thanks for testing this and your detailed steps to reproduce. I'm working on this now.

kienstra committed:
Prevent an error when removing an image

As Luke found, there was an error in the preview
view after removing an image:
"Error loading block: Invalid parameter(s): attributes"
So this sets the value to 0 on removing it.

kienstra commented:
**Addressed Bug**

Hi @lukecarbis,
Thanks for finding this bug. 

https://github.com/getblocklab/block-lab/pull/260/commits/82985b33a8f5b3f07f549c2d1f17a4d69c607371 addresses it by setting the attribute to `0` on removing the image. The back-end enforces that it's an int, so setting it to `null` didn't work.

The REST response to the `<ServerSideRender>` call had:

```
{"code":"rest_invalid_param","message":"Invalid parameter(s): attributes","data":{"status":400,"params":{"attributes":"attributes[test-image] is not of type integer."}}}
```

kienstra commented on php/blocks/controls/class-image.php:
```diff
@@ -46,15 +53,30 @@ public function register_settings() {
 		);
-		$this->settings[] = new Control_Setting(
-			array(
-				'name'     => 'default',
```
You probably saw this, but this PR removes the default image URL setting:

<img width="1117" alt="image-field" src="https://user-images.githubusercontent.com/4063887/55522135-ef0ccd00-5640-11e9-9ff0-b302348861d0.png">

This is because this control now stores the integer image ID, not the URL string.

The change below is a backward-compatibility break. It won't honor default image URLs that were entered before.

RobStino commented:
Hey fellas.
Just noticed 2 issues:
quick screen recording
https://d.pr/v/3RHw8R

kienstra commented:
**Help Text**

Hi @RobStino,
Thanks a lot for testing this.

The help text definitely isn't appearing in the [video](https://d.pr/v/3RHw8R) above.

But it seems to appear alright for me locally:

![help-text](https://user-images.githubusercontent.com/4063887/55523716-97259480-5647-11e9-908d-c0feb40ee1c6.gif)

kienstra committed:
Merge branch 'develop' into feature/image-object

kienstra commented:
**When there's no template**

The [screencast](https://d.pr/v/3RHw8R) also shows that the block displays `0 0` when no template is found. I wish I could reproduce that.

But it looks as expected locally:

![template-not-found](https://user-images.githubusercontent.com/4063887/55523920-5bd79580-5648-11e9-87b3-4b1d704a9967.gif)

kienstra commented:
I just merged in `develop`, maybe that'll help.

kienstra commented:
Maybe a fresh `npm install && npm run dev` would help if you haven't done that already.

lukecarbis commented:
Wiki updated.

Commit:
Better image placeholder

Commit:
Better styles for help text

lukecarbis commented:
@RobStino Can you please test again for help text and weird response when there's no template? The previous help text should now be replaced by a "placeholder", and help text will appear below the buttons (the position isn't easily changed).

kienstra commented on js/blocks/loader/editor.scss:
```diff
@@ -104,8 +104,7 @@ div[class^="wp-block-block-lab-"],
         vertical-align: top;
-        margin-right: 0.6rem;
-        margin-top: 0.6rem;
+        margin: 1rem 0.6rem 0.6rem 0;
```
`1rem` of margin-top might be a little too much in the Inspector. Though even the `0.6rem` that I added before was probably too much.

<img width="501" alt="space-above" src="https://user-images.githubusercontent.com/4063887/55531276-d7483f80-5666-11e9-9b8c-ef5d5ec25198.png">

kienstra commented:
**Looks Good, One Minor Comment**

Hi @lukecarbis,
The latest 2 commits look good. There's one minor comment above.

The placeholder looks really good, and prevents the block from getting shorter on removing an image:
<img width="1054" alt="placeholder" src="https://user-images.githubusercontent.com/4063887/55531481-b9c7a580-5667-11e9-81c8-4c1d267312ac.png">

RobStino commented:
The placeholder square definitely makes me think "drag and drop"...

kienstra committed:
Allow dropping images into the Image control

Use the DropZone, like how the MediaPlaceholder
component does.
Slightly adjust the styling for it.

kienstra committed:
Adjust the styling of the image buttons

Use padding on the container instead of
margins for the individual images.

kienstra commented:
**Dropzone Added**

Hi @RobStino and @lukecarbis,
https://github.com/getblocklab/block-lab/pull/260/commits/dc529cedb6a157b7c8aceacd3bfeb07a9f2371f6 allows dragging and dropping an image into the placeholder:

![drop-zone](https://user-images.githubusercontent.com/4063887/55597166-69088900-5709-11e9-8c42-f0665c273a28.gif)

And https://github.com/getblocklab/block-lab/pull/260/commits/7da2e98d55d9acf82960462721526e8e60789bcf makes a slight styling change:
<img width="444" alt="less-space" src="https://user-images.githubusercontent.com/4063887/55597121-47a79d00-5709-11e9-8daf-583ff7fda30c.png">

kienstra commented on js/blocks/controls/image.js:
```diff
@@ -68,15 +71,35 @@ const BlockLabImageControl = ( props, field, block ) => {
 			setAttributes( attr );
 		}
 
+		const uploadFiles = ( files ) => {
```
This is copied verbatim [from below](https://github.com/getblocklab/block-lab/pull/260/commits/dc529cedb6a157b7c8aceacd3bfeb07a9f2371f6#diff-7bbb33042f7b94c527dc6a5bdb42c258L94).

kienstra committed:
Remove needless argument to uploadStart()

It actually doesn't accept an argument,
and of course doesn't do anything with one.

RobStino commented:
@kienstra Great functionality on the drag-n-drop.
It just needs some some styling .
Let me demo...
![image](https://user-images.githubusercontent.com/1578148/55597785-a82eec80-5792-11e9-8ac6-0d15e46cf7b5.png)

I just changed the border to `border: 3px dashed #e2e4e7;`

The icon will just need to be better positioned.

kienstra committed:
Apply Rob's styling for the border

This makes it a little wider,
and uses a dash.

kienstra commented:
Sure, https://github.com/getblocklab/block-lab/pull/260/commits/8d04908ec24178e518c1b7b9e9765cd5e783206f applies your styling. I'm working on centering the placeholder.

kienstra committed:
Center the dashicon in the placeholder

As Rob noted,
it wasn't centered before.

kienstra commented:
<img width="870" alt="image-control" src="https://user-images.githubusercontent.com/4063887/55601729-7a5c9000-571f-11e9-89d1-9621a8f521e1.png">

kienstra commented:
@RobStino, thanks for testing this. The last 2 commits apply your suggestions.

RobStino commented:
Awesome!
I did spot something that looks like it's come over from the native Gutenberg drag-drop component. Some white text appears when you drag the image in to the drop zone.
See this video for an example.
https://d.pr/v/q5j5iJ

kienstra commented:
**Label Text**

Hi @RobStino,
Was this the text (circled in green below)?
>Drop files to upload

<img width="1396" alt="label-here" src="https://user-images.githubusercontent.com/4063887/55606277-6cb20500-5735-11e9-9364-36ef3b078596.png">

I could remove that if needed by passing `label=" "` to `<DropZone>`

But how about changing it to:
>Drop file to upload

This would change 'files' to 'file', as you can only upload one file.

lukecarbis commented:
@kienstra Nice work on the drop zone! I'll try to take a pass at this today and see if I can find something that looks good visually.

kienstra commented:
Thanks, @lukecarbis!

Commit:
Merge branch 'develop' into feature/image-object

Commit:
Add placeholder to image dropzone and tidy up styles

kienstra commented:
**Looks Good**

Hi @lukecarbis,
The [last 2 commits](https://github.com/getblocklab/block-lab/pull/260/files/e8d87fb7ec3cfd7f0a54b1f1f80a5df2a7f3c5de..60cb8c0c369fb0a28b004687e9724d4bf1c2d4f7) look good, and they worked locally. I like the help text as it is with your commits, above the image:

<img width="853" alt="help-text-could-clutter" src="https://user-images.githubusercontent.com/4063887/55764616-f877aa00-5a31-11e9-8e99-c04516ebac9b.png">

RobStino reviewed (dismissed):
Tested and looks great to me!

lukecarbis commented:
Great! I'll wait for #276 to be finished and merged, then I'll merge develop, then I'll merge this PR (but only if Travis is happy).

kienstra commented:
Haha, sounds good.

Commit:
Merge branch 'develop' into feature/image-object

Commit:
Minor restructure of image field validation method to make it PHP 5.4 compatible

lukecarbis commented:
@kienstra Could you please have a quick look at 7cda2c8 to make sure it makes sense?

kienstra commented on php/blocks/controls/class-image.php:
```diff
@@ -63,8 +63,10 @@ public function register_settings() {
 	 * @return string|int $value The value to be made available or echoed on the front-end template, possibly 0 if none found.
 	 */
 	public function validate( $value, $echo ) {
+		$image_id = intval( $value );
```
Hi @lukecarbis, this looks good. I haven't tested it, but if you're happy, I'm happy.

kienstra replied:
Good idea to set up Travis, it's good that it caught this.

lukecarbis committed:
Add unit test for returning the attachment ID

lukecarbis committed:
Fix unit test assertion

kienstra commented:
**Backwards Compatibility**

This PR has a backwards-compatibility break with `block_value( 'foo-url' )`, but not `block_field( 'foo-url' )`

`block_value( 'foo-url' )` now returns an `integer` attachment ID or `0`, as shown in the top 2 rows. 

| Function call | Expected result  |  Notes  | 
|---|---|---|
| `block_value( 'foo-valid-url' )`       | returns attachment `ID`   | Different from before, used to return the URL  |
| `block_value( 'foo-invalid-url' )`       | returns `0` | Also different from before. Returns `0` if the URL isn't on the same site, or it doesn't match an existing attachment (image). |
| `block_field( 'foo-valid-or-invalid-url' )`       | echoes the string URL | No validation added in this PR,  simply echoes the string (no change from previous behavior) |   

But `block_field( 'foo-url' )` should have the same result as before, even if it used to have an external URL.

And if the URL was previously for an external image, not an image hosted in the Media Library, `block_value( 'foo-url' )` will now return `0`.

AndrewSheetMetal commented:
Hey guys, just a quick question:
We need this feature for a project. When can we expect this feature to be distributed in a release, so we get it when we update the plugin via wordpress?
Can I otherwise download the plugin here from github?
Thanks!

kienstra commented:
**When are we releasing this?**

Hi @AndrewSheetMetal,
Sure, we're planning on releasing it this coming week. and you could download it from the [WordPress.org repo](https://wordpress.org/plugins/block-lab/). If you could, it'd be good to wait until then to download it. If it's for non-production use, you could checkout the `develop` branch from this GitHub repo, and follow the [Local Setup](https://github.com/getblocklab/block-lab/blob/develop/CONTRIBUTING.md#local-setup) steps.

AndrewSheetMetal commented:
Hey @kienstra,
thanks for the fast reply! I'm going to wait for the release then.
